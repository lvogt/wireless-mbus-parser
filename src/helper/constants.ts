export const DATA_LINK_LAYER_SIZE = 10;

export const FRAME_A_BLOCK_SIZE = 16;
export const FRAME_B_BLOCK_SIZE = 128;
