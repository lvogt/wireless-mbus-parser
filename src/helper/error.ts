export type ErrorName =
  | "CRC_ERROR"
  | "NO_AES_KEY"
  | "WRONG_AES_KEY"
  | "DECRYPTION_ERROR"
  | "DATA_RECORD_CACHE_MISSING"
  | "UNEXPECTED_STATE"
  | "UNIMPLEMENTED_FEATURE";

export class ParserError extends Error {
  name: ErrorName;
  message: string;

  constructor(name: ErrorName, message: string) {
    super();
    this.name = name;
    this.message = message;
  }
}
