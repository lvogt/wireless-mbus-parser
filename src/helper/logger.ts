export const log = {
  debug:
    process.env.DEBUG === "1"
      ? (msg: string, ...optionalParams: unknown[]) => {
          console.log(msg, ...optionalParams);
        }
      : // eslint-disable-next-line @typescript-eslint/no-empty-function
        (_msg: string, ..._optionalParams: unknown[]) => {},
  error: (msg: string, ...optionalParams: unknown[]) => {
    console.error(msg, ...optionalParams);
  },
};
