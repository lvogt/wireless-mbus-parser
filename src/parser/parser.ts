import { stripAnyCrc } from "@/crc/crcHandler";
import { ParserError } from "@/helper/error";
import { getMeterData, isCompactFrame } from "@/helper/helper";
import { decodeApplicationLayer } from "@/parser/applicationLayer";
import { decodeAuthenticationAndFragmentationLayer } from "@/parser/authenticationFragmentationLayer";
import {
  calcHeaderCrc,
  decodeDataRecords,
  extractDataRecordHeaders,
} from "@/parser/dataRecords";
import { evaluateDataRecords } from "@/parser/evaluatedData";
import { decodeExtendedLinkLayer } from "@/parser/extendedLinkLayer";
import { decodeLinkLayer } from "@/parser/linkLayer";
import type {
  ApplicationLayer,
  DataRecordHeader,
  FullParserResult,
  ParserResult,
  ParserState,
} from "@/types";

interface ParserOptions {
  key?: Buffer;
  containsCrc?: boolean;
}

export class WirelessMbusParser {
  private dataRecordHeaderCache: Record<number, DataRecordHeader[] | null> = {};

  async parseFullResult(
    data: Buffer,
    options?: Partial<ParserOptions>
  ): Promise<FullParserResult> {
    const crcFreeData = stripAnyCrc(data, options?.containsCrc);

    const state = {
      data: crcFreeData,
      pos: 0,
      key: options?.key ?? undefined,
    };

    const { state: llState, linkLayer: llFromLinkLayer } =
      decodeLinkLayer(state);

    const { state: ellState, extendedLinkLayer } = decodeExtendedLinkLayer(
      llState,
      llFromLinkLayer
    );

    const { state: aflState, authenticationAndFragmentationLayer } =
      decodeAuthenticationAndFragmentationLayer(ellState);

    const {
      state: aplState,
      applicationLayer,
      linkLayer,
    } = await decodeApplicationLayer(
      aflState,
      llFromLinkLayer,
      authenticationAndFragmentationLayer
    );
    const meterData = getMeterData(linkLayer, applicationLayer);

    const dataRecords = this.handleDataRecordDecoding(
      aplState,
      applicationLayer
    );

    const evaluatedData = evaluateDataRecords(dataRecords, meterData);

    return {
      data: evaluatedData,
      meter: meterData,
      linkLayer,
      extendedLinkLayer,
      authenticationAndFragmentationLayer,
      applicationLayer,
      dataRecords,
      rawData: aplState.data,
    };
  }

  async parse(
    data: Buffer,
    options?: Partial<ParserOptions>
  ): Promise<ParserResult> {
    const { data: evaluatedData, meter } = await this.parseFullResult(
      data,
      options
    );
    return { data: evaluatedData, meter };
  }

  private handleDataRecordDecoding(
    state: ParserState,
    applicationLayer: ApplicationLayer
  ) {
    if (isCompactFrame(applicationLayer)) {
      const dataRecordHeaders =
        this.dataRecordHeaderCache[applicationLayer.headerCrc];
      if (dataRecordHeaders === undefined || dataRecordHeaders === null) {
        this.dataRecordHeaderCache[applicationLayer.headerCrc] = null;
        throw new ParserError(
          "DATA_RECORD_CACHE_MISSING",
          "Compact frame received but data record cache is missing"
        );
      }
      const { dataRecords } = decodeDataRecords(state, dataRecordHeaders);
      return dataRecords;
    } else {
      const { dataRecords } = decodeDataRecords(state);

      const headerCrc = calcHeaderCrc(dataRecords, state.data);
      if (this.dataRecordHeaderCache[headerCrc] === null) {
        this.dataRecordHeaderCache[headerCrc] =
          extractDataRecordHeaders(dataRecords);
      }

      return dataRecords;
    }
  }
}
