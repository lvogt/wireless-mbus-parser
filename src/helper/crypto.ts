import {
  type CipherGCMTypes,
  type CipherOCBTypes,
  createDecipheriv,
} from "crypto";

import { log } from "@/helper/logger";

export function decryptData(
  encryptedData: Buffer,
  iv: Buffer,
  key: Buffer,
  algorithm: CipherGCMTypes | CipherGCMTypes | CipherOCBTypes | string
) {
  const padding = encryptedData.length % 16;
  const length = encryptedData.length;

  const decipher = createDecipheriv(algorithm, key, iv);
  decipher.setAutoPadding(false);

  if (padding) {
    encryptedData = Buffer.concat([encryptedData, Buffer.alloc(16 - padding)]);
  }

  const decryptedData = Buffer.concat([
    decipher.update(encryptedData),
    decipher.final(),
  ]);

  return decryptedData.subarray(0, length);
}

export function decryptInPlace(
  data: Buffer,
  key: Buffer,
  iv: Buffer,
  algorithm: string,
  offset: number,
  length: number
) {
  const encryptedData = data.subarray(offset, offset + length);
  log.debug(`Encrypted data: ${encryptedData.toString("hex")}`);
  log.debug(`IV: ${iv.toString("hex")}`);
  const decryptedData = decryptData(encryptedData, iv, key, algorithm);

  log.debug(`Decrypted data: ${decryptedData.toString("hex")}`);
  if (!decryptedData.length) {
    throw new Error(
      `Decryption (${algorithm}) failed! IV: ${iv.toString("hex")}`
    );
  }

  return Buffer.concat([
    data.subarray(0, offset),
    decryptedData,
    data.subarray(offset + length),
  ]);
}
