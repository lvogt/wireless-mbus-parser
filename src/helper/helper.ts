import { CI_RESP_COMPACT, VALID_DEVICES_TYPES } from "@/helper/constants";
import { log } from "@/helper/logger";
import type {
  ApplicationLayer,
  ApplicationLayerCompact,
  LinkLayer,
  MeterData,
  PrimaryVif,
  PrimaryVifString,
  WiredLinkLayer,
} from "@/types";

export function isWiredMbusFrame(data: Buffer) {
  return data[0] == 0x68 && data[3] == 0x68 && data[data.length - 1] == 0x16;
}

export function isLinkLayer(ll: LinkLayer | WiredLinkLayer): ll is LinkLayer {
  return (ll as LinkLayer).mField !== undefined;
}

export function isCompactFrame(
  apl: ApplicationLayer
): apl is ApplicationLayerCompact {
  return (apl as ApplicationLayerCompact).ci === CI_RESP_COMPACT;
}

export function isPrimaryVifString(vif: PrimaryVif): vif is PrimaryVifString {
  return (vif as PrimaryVifString).plainText !== undefined;
}

export function getMeterData(
  linkLayer: LinkLayer,
  applicationLayer?: ApplicationLayer
): MeterData {
  const linkLayerMeterData = {
    manufacturer: linkLayer.manufacturer,
    id: linkLayer.meterId,
    type: linkLayer.type,
    deviceType: getDeviceType(linkLayer.type),
    version: linkLayer.version,
  };

  if (applicationLayer?.ci == 0x7a) {
    return {
      ...linkLayerMeterData,
      status: applicationLayer.status,
      accessNo: applicationLayer.accessNo,
    };
  } else if (applicationLayer?.ci === 0x72) {
    return {
      manufacturer: applicationLayer.meterManufacturerString,
      id: applicationLayer.meterIdString,
      type: applicationLayer.meterDevice,
      deviceType: getDeviceType(applicationLayer.meterDevice),
      version: applicationLayer.meterVersion,
      status: applicationLayer.status,
      accessNo: applicationLayer.accessNo,
      radio: linkLayerMeterData,
    };
  } else {
    return linkLayerMeterData;
  }
}

export function decodeBCD(digits: number, data: Buffer, offset = 0) {
  const length = Math.ceil(digits / 2);
  const bcdData = Buffer.from(data.subarray(offset, offset + length));

  // check for negative BCD
  const sign = bcdData[length - 1] >> 4 > 9 ? -1 : 1;
  if (sign === -1) {
    bcdData[length - 1] &= 0x0f;
  }

  let val = 0;
  for (let i = 0; i < length; i++) {
    val +=
      ((bcdData[i] & 0x0f) + ((bcdData[i] & 0xf0) >> 4) * 10) *
      Math.pow(100, i);
  }

  return Math.trunc(sign * val);
}

export function getMeterId(data: Buffer | number, offset = 0) {
  const num = Buffer.isBuffer(data) ? data.readUint32LE(offset) : data;
  return num.toString(16).padStart(8, "0");
}

export function decodeManufacturer(data: number) {
  return (
    String.fromCharCode((data >> 10) + 64) +
    String.fromCharCode(((data >> 5) & 0x1f) + 64) +
    String.fromCharCode((data & 0x1f) + 64)
  );
}

export function getDeviceType(type: number) {
  const idx = type as keyof typeof VALID_DEVICES_TYPES;
  return VALID_DEVICES_TYPES[idx] || "unknown";
}

export function getDeviceState(statusCode: number) {
  const lowPower = statusCode & 0b00000100;
  const permanentError = statusCode & 0b00001000;
  const temporaryError = statusCode & 0b00010000;
  const manufacturerSpecific = statusCode & 0b11100000;

  const suffix =
    (lowPower ? "; low power" : "") +
    (permanentError ? " (permanent)" : "") +
    (temporaryError ? " (temporary)" : "") +
    (manufacturerSpecific
      ? "; Manufacturer specific 0b" +
        manufacturerSpecific.toString(2).padStart(8, "0")
      : "");

  const bit01 = statusCode & 0b00000011;
  switch (bit01) {
    case 0x00:
      return "No error" + suffix;
    case 0x01:
      return "Busy" + suffix;
    case 0x02:
      return "Error" + suffix;
    case 0x03:
      return "Alarm" + suffix;
  }
}

// this assumes the telegram contains a local time and
// parses it with respect to the current timezone
export function decodeDateTimeTypeF(value: number) {
  // value is a 32bit int
  // z = daylight saving time?; v = valid?; r = reserved
  //   YYYY MMMM YYYD DDDD zrrh hhhh vrmm mmmm
  // 0b0001 0101 0001 1111 0011 0111 0011 0010 = 2008-05-31 23:50

  const valid = (value & 0x80) >> 8 === 0;
  if (!valid) {
    log.debug("Warning: Invalid date encountered!");
  }

  const minutes = value & 0x3f;
  const hours = (value & 0x1f00) >> 8;
  const day = (value & 0x1f0000) >> 16;
  const month = (value & 0x0f000000) >> 24;
  const year = ((value & 0xf0000000) >> 25) | ((value & 0xe00000) >> 21);
  return new Date(year + 2000, month - 1, day, hours, minutes, 0);
}

export function decodeDateTypeG(value: number) {
  // value is a 16bit int
  //   YYYY MMMM YYYD DDDD
  // 0b0000 1100 1111 1111 = 31.12.2007
  // 0b0000 0100 1111 1110 = 30.04.2007

  const day = value & 0x1f;
  const month = (value & 0x0f00) >> 8;
  const year = ((value & 0xf000) >> 9) | ((value & 0xe0) >> 5);
  return new Date(year + 2000, month - 1, day);
}

export function decodeDateTimeTypeI(value: number) {
  // value is a 48bit int - msb = 0x00 / unused
  //   YYYY MMMM YYYD DDDD WWWh hhhh 00mm mmmm 00ss ssss
  // 0b0010 0001 1110 1010 0100 1110 0001 0001 0011 1000 = 2023-01-10 14:17:56

  const buffer = Buffer.alloc(6);
  buffer.writeUIntLE(value, 0, 6);

  const seconds = buffer[0] & 0x3f;
  const minutes = buffer[1] & 0x3f;
  const hours = buffer[2] & 0x1f;
  const day = buffer[3] & 0x1f;
  const month = buffer[4] & 0x0f;
  const year = ((buffer[3] & 0xe0) >> 5) | ((buffer[4] & 0xf0) >> 1);
  return new Date(year + 2000, month - 1, day, hours, minutes, seconds);
}

export function decodeDateTypeTechem(value: number) {
  // value is a 16bit int

  //   YYYY YYYM MMMD DDDD
  // 0b0010 0101 1001 1111 = 31.12.2018

  const day = value & 0x1f;
  const month = (value & 0x1e0) >> 5;
  const year = (value & 0x7e00) >> 9;
  return new Date(year + 2000, month - 1, day);
}

export function decodeNoYearDateTypeTechem(value: number) {
  // value is a 16bit int
  //   rrrM MMMD DDDD rrrr
  // 0b0000 1100 1111 0000 = 15.06.

  const day = (value & 0x1f0) >> 4;
  const month = (value & 0x1e00) >> 9;
  const year = new Date().getFullYear();
  return new Date(year, month - 1, day);
}

export function decodeNoYearDateType2Techem(
  valueDay: number,
  valueMonth: number
) {
  const day = (valueDay & 0xf80) >> 7;
  const month = (valueMonth & 0x78) >> 3;
  const year = new Date().getFullYear();
  return new Date(year, month - 1, day);
}

export function encodeDateTypeG(date: Date) {
  const day = date.getDate();
  const month = date.getMonth() + 1;
  const year = date.getFullYear() - 2000;

  const yearLSB = year & 0b111;
  const yearMSB = (year & 0b1111000) >> 3;

  //   YYYY MMMM YYYD DDDD
  // 0b0000 1100 1111 1111 = 31.12.2007
  // 0b0000 0100 1111 1110 = 30.04.2007

  return (yearMSB << 12) | (month << 8) | (yearLSB << 5) | day;
}
