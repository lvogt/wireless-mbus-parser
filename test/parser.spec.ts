import { fail } from "assert";
import { describe, expect, it } from "vitest";

import { ParserError } from "@/helper/error";
import { WirelessMbusParser } from "@/parser/parser";

async function decode(data: string, key?: string) {
  const parser = new WirelessMbusParser();
  return await parser.parse(
    Buffer.from(data, "hex"),
    key ? { key: Buffer.from(key, "hex") } : undefined
  );
}

async function decodeWithError(
  data: string,
  name: string,
  message: string,
  key?: string
) {
  let result;
  try {
    result = await decode(data, key);
  } catch (e: unknown) {
    if (e instanceof ParserError) {
      expect(e.name).toEqual(name);
      expect(e.message).toEqual(message);
      return;
    } else {
      fail("Error should be of type ParserError!");
    }
  }
  console.log(JSON.stringify(result));
  fail("Did not throw!");
}

describe("Parser", () => {
  it("Simple telegram", async () => {
    const result = await decode(
      "2C44A7320613996707047A821000202F2F0C06000000000C14000000000C22224101000B5A4102000B5E4000F0"
    );

    expect(result.data).toHaveLength(5);
  });
});

describe("Errors", () => {
  it("ELL: Encrypted but no key", async () => {
    await decodeWithError(
      "24442D2C692845631B168D3050209CD621B006B1140AEF4953AE5B86FAFC0B00E70705B846",
      "NO_AES_KEY",
      "Encrytped ELL, but no key provided!"
    );
  });

  it("ELL: Encrypted but wrong key", async () => {
    await decodeWithError(
      "24442D2C692845631B168D3050209CD621B006B1140AEF4953AE5B86FAFC0B00E70705B846",
      "WRONG_AES_KEY",
      "Payload CRC check failed on ExtendedLinkLayer, wrong AES key?",
      "4E5508544202058100DFEFA06B0934AF"
    );
  });

  it("Mode5+7: Encrypted but no key", async () => {
    await decodeWithError(
      "2E44931578563412330333637A2A0020255923C95AAA26D1B2E7493BC2AD013EC4A6F6D3529B520EDFF0EA6DEFC955B29D6D69EBF3EC8A",
      "NO_AES_KEY",
      "Encrypted telegram but no AES key provided"
    );
  });

  it("Mode5: Encrypted but wrong key", async () => {
    await decodeWithError(
      "2E44931578563412330333637A2A0020255923C95AAA26D1B2E7493BC2AD013EC4A6F6D3529B520EDFF0EA6DEFC955B29D6D69EBF3EC8A",
      "WRONG_AES_KEY",
      "Decryption failed, wrong key?",
      "4E5508544202058100DFEFA06B0934AF"
    );
  });

  it("Mode7: Encrypted but wrong key", async () => {
    await decodeWithError(
      "434493157856341233037AC98C2075900F002C25B30A000021924D4FBA372FB66E017A75002007109058475F4BC9D1281DF878B80A1B0F98B629024AAC7279429398BFC549233C0140829B93BAA1",
      "WRONG_AES_KEY",
      "Received MAC does not match. Wrong key?\nMAC received: 21924d4f2fb66e01",
      "4E5508544202058100DFEFA06B0934AF"
    );
  });

  it("Compact frame without cache", async () => {
    await decodeWithError(
      "3F442D2C06357260190C8D207C71032F21255C79DD829283011117650000BFA80000D24F0000B1FB00000000E919FF18F7640000E8FA00000B000000DB111C0B",
      "DATA_RECORD_CACHE_MISSING",
      "Compact frame received but data record cache is missing"
    );
  });

  it("Compact frame without cache, afterwards with", async () => {
    const parser = new WirelessMbusParser();
    try {
      await parser.parse(
        Buffer.from(
          "3F442D2C06357260190C8D207C71032F21255C79DD829283011117650000BFA80000D24F0000B1FB00000000E919FF18F7640000E8FA00000B000000DB111C0B",
          "hex"
        )
      );
    } catch {
      /* empty */
    }

    expect(parser["dataRecordHeaderCache"]).toEqual({ 0x82dd: null });

    await parser.parse(
      Buffer.from(
        "5C442D2C06357260190C8D207B70032F21271D7802F9FF15011104061765000004EEFF07BFA8000004EEFF08D24F00000414B1FB000002FD170000026CE919426CFF184406F76400004414E8FA0000043B0B0000000259DB11025D1C0B",
        "hex"
      )
    );

    const result = await parser.parse(
      Buffer.from(
        "3F442D2C06357260190C8D207C71032F21255C79DD829283011117650000BFA80000D24F0000B1FB00000000E919FF18F7640000E8FA00000B000000DB111C0B",
        "hex"
      )
    );

    expect(result.data).toHaveLength(13);
    expect(parser["dataRecordHeaderCache"][0x82dd]).not.toEqual(null);
  });

  it("Cache only populated if needed", async () => {
    const parser = new WirelessMbusParser();
    await parser.parse(
      Buffer.from(
        "5C442D2C06357260190C8D207B70032F21271D7802F9FF15011104061765000004EEFF07BFA8000004EEFF08D24F00000414B1FB000002FD170000026CE919426CFF184406F76400004414E8FA0000043B0B0000000259DB11025D1C0B",
        "hex"
      )
    );

    expect(parser["dataRecordHeaderCache"]).toEqual({});
  });
});
