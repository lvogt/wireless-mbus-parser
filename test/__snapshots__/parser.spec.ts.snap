// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Parser > All fields 1`] = `
{
  "applicationLayer": {
    "accessNo": 117,
    "ci": 122,
    "config": {
      "content": 0,
      "encryptedBlocks": 2,
      "kdfSel": 1,
      "keyid": 0,
      "mode": 7,
    },
    "offset": 30,
    "status": "No error",
    "statusCode": 0,
  },
  "authenticationAndFragmentationLayer": {
    "afll": 15,
    "ci": 144,
    "fcl": {
      "fid": 0,
      "kip": false,
      "macp": true,
      "mclp": true,
      "mcrp": true,
      "mf": false,
      "mlp": false,
    },
    "fclRaw": 11264,
    "ki": undefined,
    "kiRaw": undefined,
    "mac": {
      "data": [
        33,
        146,
        77,
        79,
        47,
        182,
        110,
        1,
      ],
      "type": "Buffer",
    },
    "mcl": {
      "at": 5,
      "kimp": false,
      "mcmp": true,
      "mlmp": false,
    },
    "mclRaw": 37,
    "mcr": 2739,
    "ml": undefined,
  },
  "data": [
    {
      "description": "Volume",
      "info": {
        "deviceUnit": 0,
        "legacyVif": "VIF_VOLUME",
        "storageNo": 0,
        "tariff": 0,
      },
      "type": 0,
      "unit": "m³",
      "value": 28504.27,
    },
    {
      "description": "Time point",
      "info": {
        "deviceUnit": 0,
        "legacyVif": "VIF_TIME_POINT_DATE_TIME",
        "storageNo": 0,
        "tariff": 0,
      },
      "type": 4,
      "unit": "",
      "value": 2008-05-31T23:50:00.000Z,
    },
    {
      "description": "Error flags (binary)",
      "info": {
        "deviceUnit": 0,
        "legacyVif": "VIF_ERROR_FLAGS",
        "storageNo": 0,
        "tariff": 0,
      },
      "type": 0,
      "unit": "",
      "value": 0,
    },
  ],
  "dataRecords": [
    {
      "header": {
        "dib": {
          "dataField": 12,
          "deviceUnit": 0,
          "functionField": 0,
          "storageNo": 0,
          "tariff": 0,
        },
        "length": 2,
        "offset": 38,
        "vib": {
          "extensions": [],
          "primary": {
            "extensionBitSet": false,
            "table": 0,
            "vif": 20,
          },
        },
      },
      "value": 2850427,
    },
    {
      "header": {
        "dib": {
          "dataField": 4,
          "deviceUnit": 0,
          "functionField": 0,
          "storageNo": 0,
          "tariff": 0,
        },
        "length": 2,
        "offset": 44,
        "vib": {
          "extensions": [],
          "primary": {
            "extensionBitSet": false,
            "table": 0,
            "vif": 109,
          },
        },
      },
      "value": 354367282,
    },
    {
      "header": {
        "dib": {
          "dataField": 2,
          "deviceUnit": 0,
          "functionField": 0,
          "storageNo": 0,
          "tariff": 0,
        },
        "length": 3,
        "offset": 50,
        "vib": {
          "extensions": [],
          "primary": {
            "extensionBitSet": false,
            "table": 1,
            "vif": 23,
          },
        },
      },
      "value": 0,
    },
  ],
  "extendedLinkLayer": {
    "accessNumber": 117,
    "ci": 140,
    "communicationControl": 32,
  },
  "linkLayer": {
    "aField": 305419896,
    "aFieldRaw": {
      "data": [
        120,
        86,
        52,
        18,
        51,
        3,
      ],
      "type": "Buffer",
    },
    "addressRaw": {
      "data": [
        147,
        21,
        120,
        86,
        52,
        18,
        51,
        3,
      ],
      "type": "Buffer",
    },
    "cField": 68,
    "lField": 67,
    "mField": 5523,
    "manufacturer": "ELS",
    "meterId": "12345678",
    "type": 3,
    "typeString": "Gas",
    "version": 51,
  },
  "meter": {
    "accessNo": 117,
    "deviceType": "Gas",
    "id": "12345678",
    "manufacturer": "ELS",
    "status": "No error",
    "type": 3,
    "version": 51,
  },
  "rawData": {
    "data": [
      67,
      68,
      147,
      21,
      120,
      86,
      52,
      18,
      51,
      3,
      140,
      32,
      117,
      144,
      15,
      0,
      44,
      37,
      179,
      10,
      0,
      0,
      33,
      146,
      77,
      79,
      47,
      182,
      110,
      1,
      122,
      117,
      0,
      32,
      7,
      16,
      47,
      47,
      12,
      20,
      39,
      4,
      133,
      2,
      4,
      109,
      50,
      55,
      31,
      21,
      2,
      253,
      23,
      0,
      0,
      47,
      47,
      47,
      47,
      47,
      47,
      47,
      47,
      47,
      47,
      47,
      47,
      47,
    ],
    "type": "Buffer",
  },
}
`;

exports[`Parser > Full result 1`] = `
{
  "applicationLayer": {
    "accessNo": 130,
    "ci": 122,
    "config": {
      "accessability": false,
      "bidirectional": false,
      "content": 0,
      "encryptedBlocks": 0,
      "hopCounter": 0,
      "mode": 0,
      "synchronous": true,
    },
    "offset": 10,
    "status": "No error (temporary)",
    "statusCode": 16,
  },
  "authenticationAndFragmentationLayer": undefined,
  "data": [
    {
      "description": "Energy",
      "info": {
        "deviceUnit": 0,
        "legacyVif": "VIF_ENERGY_WATT",
        "storageNo": 0,
        "tariff": 0,
      },
      "type": 0,
      "unit": "Wh",
      "value": 0,
    },
    {
      "description": "Volume",
      "info": {
        "deviceUnit": 0,
        "legacyVif": "VIF_VOLUME",
        "storageNo": 0,
        "tariff": 0,
      },
      "type": 0,
      "unit": "m³",
      "value": 0,
    },
    {
      "description": "On Time",
      "info": {
        "deviceUnit": 0,
        "legacyVif": "VIF_ON_TIME",
        "storageNo": 0,
        "tariff": 0,
      },
      "type": 0,
      "unit": "h",
      "value": 14122,
    },
    {
      "description": "Flow Temperature",
      "info": {
        "deviceUnit": 0,
        "legacyVif": "VIF_FLOW_TEMP",
        "storageNo": 0,
        "tariff": 0,
      },
      "type": 0,
      "unit": "°C",
      "value": 24.1,
    },
    {
      "description": "Return Temperature",
      "info": {
        "deviceUnit": 0,
        "legacyVif": "VIF_RETURN_TEMP",
        "storageNo": 0,
        "tariff": 0,
      },
      "type": 0,
      "unit": "°C",
      "value": -4,
    },
  ],
  "dataRecords": [
    {
      "header": {
        "dib": {
          "dataField": 12,
          "deviceUnit": 0,
          "functionField": 0,
          "storageNo": 0,
          "tariff": 0,
        },
        "length": 2,
        "offset": 17,
        "vib": {
          "extensions": [],
          "primary": {
            "extensionBitSet": false,
            "table": 0,
            "vif": 6,
          },
        },
      },
      "value": 0,
    },
    {
      "header": {
        "dib": {
          "dataField": 12,
          "deviceUnit": 0,
          "functionField": 0,
          "storageNo": 0,
          "tariff": 0,
        },
        "length": 2,
        "offset": 23,
        "vib": {
          "extensions": [],
          "primary": {
            "extensionBitSet": false,
            "table": 0,
            "vif": 20,
          },
        },
      },
      "value": 0,
    },
    {
      "header": {
        "dib": {
          "dataField": 12,
          "deviceUnit": 0,
          "functionField": 0,
          "storageNo": 0,
          "tariff": 0,
        },
        "length": 2,
        "offset": 29,
        "vib": {
          "extensions": [],
          "primary": {
            "extensionBitSet": false,
            "table": 0,
            "vif": 34,
          },
        },
      },
      "value": 14122,
    },
    {
      "header": {
        "dib": {
          "dataField": 11,
          "deviceUnit": 0,
          "functionField": 0,
          "storageNo": 0,
          "tariff": 0,
        },
        "length": 2,
        "offset": 35,
        "vib": {
          "extensions": [],
          "primary": {
            "extensionBitSet": false,
            "table": 0,
            "vif": 90,
          },
        },
      },
      "value": 241,
    },
    {
      "header": {
        "dib": {
          "dataField": 11,
          "deviceUnit": 0,
          "functionField": 0,
          "storageNo": 0,
          "tariff": 0,
        },
        "length": 2,
        "offset": 40,
        "vib": {
          "extensions": [],
          "primary": {
            "extensionBitSet": false,
            "table": 0,
            "vif": 94,
          },
        },
      },
      "value": -40,
    },
  ],
  "extendedLinkLayer": undefined,
  "linkLayer": {
    "aField": 1738085126,
    "aFieldRaw": {
      "data": [
        6,
        19,
        153,
        103,
        7,
        4,
      ],
      "type": "Buffer",
    },
    "addressRaw": {
      "data": [
        167,
        50,
        6,
        19,
        153,
        103,
        7,
        4,
      ],
      "type": "Buffer",
    },
    "cField": 68,
    "lField": 44,
    "mField": 12967,
    "manufacturer": "LUG",
    "meterId": "67991306",
    "type": 4,
    "typeString": "Heat",
    "version": 7,
  },
  "meter": {
    "accessNo": 130,
    "deviceType": "Heat",
    "id": "67991306",
    "manufacturer": "LUG",
    "status": "No error (temporary)",
    "type": 4,
    "version": 7,
  },
  "rawData": {
    "data": [
      44,
      68,
      167,
      50,
      6,
      19,
      153,
      103,
      7,
      4,
      122,
      130,
      16,
      0,
      32,
      47,
      47,
      12,
      6,
      0,
      0,
      0,
      0,
      12,
      20,
      0,
      0,
      0,
      0,
      12,
      34,
      34,
      65,
      1,
      0,
      11,
      90,
      65,
      2,
      0,
      11,
      94,
      64,
      0,
      240,
    ],
    "type": "Buffer",
  },
}
`;
